Search result filtering

SamiHiltunen opened the pull request:
Adds support for filtering results of the search. Addresses a use case where I want to search for objects in the tree in a specified area with specified characteristics with a specified amount of results. Doing the filtering while searching prevents problems like asking for 100 results but after filtering the returned results, only having 45 viable results at hand.

Also removes documentation about SearchContained function which seems to be missing.

dhconnelly commented:
Hi, thanks for this! Looking now.

dhconnelly commented on README.md:
```diff
@@ -110,18 +110,26 @@ corrupt the tree.
+
+  type Filter interface {
+		Filter(results []Spatial, object Spatial) (refuse, abort bool)
+	}
```
the formatting looks a bit off here

dhconnelly commented on README.md:
```diff
@@ -110,18 +110,26 @@ corrupt the tree.
+A filter for limiting results by result count is included in the package.
+
+	// maximum of three results will be returned
+  tree.SearchIntersect(bb, NewLimitFilter(3))
```
also here

dhconnelly commented on rtree.go:
```diff
@@ -416,35 +416,31 @@ func (tree *Rtree) condenseTree(n *node) {
-// like SearchIntersect and returns all the results.
-func (tree *Rtree) SearchIntersectWithLimit(k int, bb *Rect) []Spatial {
-	return tree.searchIntersect(k, tree.root, bb)
+func (tree *Rtree) SearchIntersect(bb *Rect, filters ...Filter) []Spatial {
```
Adding a new function which accepts filters, to which the existing function delegates, would avoid breaking existing clients who aren't interested in filtering. This seems to match your intention in adding the LimitFilter to this lib (as opposed to keeping it in client libraries), but it still requires client refactoring if we don't keep the current function signature. What are your thoughts?

dhconnelly commented on filter.go:
```diff
@@ -0,0 +1,54 @@
+			abort = true
+		}
+
+		if refuse && abort {
```
can you add a comment explaining why this case requires both refuse and abort for early break?

SamiHiltunen replied:
Added a comment for this. This mainly comes down to having multiple filters in an unspecified order. If a filter aborts and ```refuse``` is not true, the leaf being filtered will still be appended to the result set. Some other filter after the aborting one could still refuse the value.

dhconnelly commented:
sorry for the delay - will look again tonight

dhconnelly commented:
Thanks for maintaining backwards compatibility -- looks great! I think the only thing that might help is another unit test for the new function, using a test filter. What are your thoughts?

SamiHiltunen commented:
Hey, I added a test with a test filter. I changed the Filter interface to be a function type which makes it much more convenient to use. Also, there are some minor changes to docs and style mostly to fix linter warnings.
